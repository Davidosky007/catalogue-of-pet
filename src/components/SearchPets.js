import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { ANIMALS } from '@frontendmasters/pet';
import PetsList from './PetList';
import petFinder from '../helpers/petFinderApi';

const SearchPets = ({
  pets = [],
  fetchPets,
  filterPets,
  filter,
  filterSize,
  size,
  gender,
  filterGender,
}) => {
  const [petSize, petSizeSet] = useState(size);
  const [petFilter, petFilterSet] = useState(filter);
  const [petGender, petGenderSet] = useState(gender || 'All');
  const [updatePets, setUpdatePets] = useState(pets || []);

  const sizes = ['Small', 'Medium', 'Large'];
  const genders = ['Male', 'Female'];
  const capitalize = ([first, ...rest]) => first.toUpperCase() + rest.join('').toLowerCase();

  async function petsRequest() {
    const { animals } = await petFinder.animal
      .search({
        type: petFilter,
        gender: petGender,
        size: petSize,
      })
      .then((data) => data.data)
      .catch((error) => error);

    fetchPets(animals);
    setUpdatePets(animals);
    filterPets(petFilter);
    filterSize(petSize);
    filterGender(petGender);
  }

  const onFilterChange = (e) => {
    const { value } = e.target;
    petFilterSet(value);
  };

  const onSizeChange = (e) => {
    const { value } = e.target;
    petSizeSet(value);
  };

  const onGenderChange = (e) => {
    const { value } = e.target;
    petGenderSet(value);
  };

  useEffect(async () => {
    const { animals } = await petFinder.animal
      .search({
        type: 'Horse',
        gender: 'Male',
        size: 'Small',
      })
      .then((data) => data.data);

    fetchPets(animals);
    setUpdatePets(animals);
  }, []);

  return (
    <div className="main__wrapper" data-testid="search-pets-id">
      <div className="search__form__wrapper">
        <form
          data-testid="search-pets-form"
          className="search__form__wrapper-form"
          onSubmit={(e) => {
            e.preventDefault();
            petsRequest();
          }}
        >
          <div className="search__form__wrapper-form-box">
            <label htmlFor="id-1">
              Animal
              <select value={petFilter} onChange={onFilterChange}>
                <option>Select</option>
                {ANIMALS.map((option) => (
                  <option value={option} key={option}>
                    {capitalize(option)}
                  </option>
                ))}
              </select>
            </label>
            <label htmlFor="id-2">
              Size
              <select value={petSize} onChange={onSizeChange}>
                <option>Select</option>
                {sizes.map((option) => (
                  <option value={option} key={option}>
                    {option}
                  </option>
                ))}
              </select>
            </label>

            <label htmlFor="id-3">
              Gender
              <select value={petGender} onChange={onGenderChange}>
                <option>Select</option>
                {genders.map((option) => (
                  <option value={option} key={option}>
                    {option}
                  </option>
                ))}
              </select>
            </label>
          </div>
          <button type="submit">Search Pet</button>
        </form>
      </div>

      <PetsList pets={updatePets} />
    </div>
  );
};

SearchPets.propTypes = {
  pets: PropTypes.array, // eslint-disable-line
  fetchPets: PropTypes.func,
  filter: PropTypes.string,
  filterPets: PropTypes.func,
  filterSize: PropTypes.func,
  filterGender: PropTypes.func,
  size: PropTypes.string,
  gender: PropTypes.string,
};

SearchPets.defaultProps = {
  pets: [],
  fetchPets: null,
  filter: 'All',
  filterPets: null,
  filterSize: null,
  filterGender: null,
  size: 'All',
  gender: 'All',
};

export default SearchPets;
