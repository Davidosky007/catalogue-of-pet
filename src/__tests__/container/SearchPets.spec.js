import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import SearchPets from '../../components/SearchPets';

test('should render Details components properly ', () => {
  const details = renderer.create(<SearchPets />);
  const final = details.toJSON();
  expect(final).toMatchSnapshot();
});

test('should render text Animal', () => {
  render(<SearchPets />);
  const details = screen.getByTestId('search-pets-id');
  expect(details).toHaveTextContent('Animal');
});
test('should render text Gender', () => {
  render(<SearchPets />);
  const details = screen.getByTestId('search-pets-id');
  expect(details).toHaveTextContent('Gender');
});
test('should render text Size', () => {
  render(<SearchPets />);
  const details = screen.getByTestId('search-pets-id');
  expect(details).toHaveTextContent('Size');
});
