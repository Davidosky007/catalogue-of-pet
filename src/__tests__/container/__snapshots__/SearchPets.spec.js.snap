// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Details components properly  1`] = `
<div
  className="main__wrapper"
  data-testid="search-pets-id"
>
  <div
    className="search__form__wrapper"
  >
    <form
      className="search__form__wrapper-form"
      data-testid="search-pets-form"
      onSubmit={[Function]}
    >
      <div
        className="search__form__wrapper-form-box"
      >
        <label
          htmlFor="id-1"
        >
          Animal
          <select
            onChange={[Function]}
            value="All"
          >
            <option>
              Select
            </option>
            <option
              value="barnyard"
            >
              Barnyard
            </option>
            <option
              value="bird"
            >
              Bird
            </option>
            <option
              value="cat"
            >
              Cat
            </option>
            <option
              value="dog"
            >
              Dog
            </option>
            <option
              value="horse"
            >
              Horse
            </option>
            <option
              value="rabbit"
            >
              Rabbit
            </option>
            <option
              value="scales-fins-other"
            >
              Scales-fins-other
            </option>
            <option
              value="small-furry"
            >
              Small-furry
            </option>
          </select>
        </label>
        <label
          htmlFor="id-2"
        >
          Size
          <select
            onChange={[Function]}
            value="All"
          >
            <option>
              Select
            </option>
            <option
              value="Small"
            >
              Small
            </option>
            <option
              value="Medium"
            >
              Medium
            </option>
            <option
              value="Large"
            >
              Large
            </option>
          </select>
        </label>
        <label
          htmlFor="id-3"
        >
          Gender
          <select
            onChange={[Function]}
            value="All"
          >
            <option>
              Select
            </option>
            <option
              value="Male"
            >
              Male
            </option>
            <option
              value="Female"
            >
              Female
            </option>
          </select>
        </label>
      </div>
      <button
        type="submit"
      >
        Search Pet
      </button>
    </form>
  </div>
  <div
    className="pets__wrapper"
    data-testid="pets-list-id"
  >
    <h1
      className="loading__wrapper"
    >
      No pets Found
    </h1>
  </div>
</div>
`;
